rep-2007.rst

```diff
@@ -0,0 +1,246 @@
+
+An additional reason for this change is performance.
+This interface will allow us to define methods for serializing directly to the user requested type, and/or using that type in intra-process communication without ever converting it.
+Whereas without this feature, even to send a ``cv::Mat`` from one publisher to a subscription in the same process would require converting it to a ``sensor_msgs::msg::Image`` first and then back again to cv::Mat.
+
+
+Terminology
+===========
+
+.. glossary::
+     Custom type (in code, ``CustomType``)
+       A data structure that *is not* a ROS 2 interface, such as ``std::string`` or ``cv::Mat``.
+     ROS type (in code, ``RosType``)
+       A data structure that *is* a ROS 2 interface, such as ``std_msgs::msg::String`` or ``sensor_msgs::msg::Image``.
+
+
+Specification
+=============
+
+There are different tiers of expressivity in the syntax for type adaptation, including both explicit to implicit forms.
+The explicit forms require the custom type and the ROS type to be specified; 
+while the implicit form makes an assumption about the ROS type based on the custom type.
+The explicit forms may be used to clearly communicate what is happening, which could be useful in examples or tutorials, or may be used depending on the preferences of the developer.
+While the implicit, and most concise, form will primarily be used for its convenience (although it lacks some of the functionality of the explicit forms).
+
+Below are examples of different syntaxes for type adaptation.
+In these examples and using the terminology above, ``std::string`` will be considered the custom type and the ROS interface ``std_msgs::msg::String`` will be considered the ROS type.
+Note that, for every publisher created in the examples below, a ``std::string`` type variable (the custom type) would be adapted to be published on the ROS network as a ``std_msgs::msg::String`` (the ROS type).
+
+.. code-block:: cpp
+
+   // Explicit Form 1: Most explicit
+   using MsgT = rclcpp::AdaptType<std::string>::to<std_msgs::msg::String>;
+   auto pub1 = std::make_shared<rclcpp::Publisher<MsgT>>();
+
+The most explicit form could be shortened in the following to ways:
+
+.. code-block:: cpp
+
+   // Explicit Form 2
+   using MsgT = rclcpp::AdaptType<std::string, std_msgs::msg::String>;
+   auto pub2 = std::make_shared<rclcpp::Publisher<MsgT>>();
+
+   // Explicit Form 3
+   auto pub3 = std::make_shared<rclcpp::Publisher<std::string, std_msgs::msg::String>>();
+
+Note that the 3rd explicit form may be prohibitively complicated to implement since there are multiple template arguments already for publishers and subscribers.
+
+The final shorthand performs an implicit type adaptation with the custom type ``std::string`` which will be adapted to the ROS type``std_msgs::msg::String``:
+
+.. code-block:: cpp
+
+   // Implicit Form
+   auto pub4 = std::make_shared<rclcpp::Publisher<std::string>>();
```

Comment:
I'm not sure this one makes sense - how and where are the adapters declared? Or, is this feature essentially declaring that a publisher declaring any ROS interface can effectively carry any arbitrary binary payload?

Reply:
> how and where are the adapters declared?

I don't believe we've seen that in this document yet. Copying from @audrow and my development prototype, this declaration looks something like this:

```c++
template<>
struct rclcpp::TypeAdaptor<std::string>
{
  using is_specialized = std::true_type;
  using ROSMessageType = std_msgs::msg::String;

  static
  ROSMessageType
  convert_to_ros_message(const std::string & input_msg)
  {
    return std_msgs::msg::String{input_msg};
  }
};
```

> Or, is this feature essentially declaring that a publisher declaring any ROS interface can effectively carry any arbitrary binary payload?

No, it's coincidence that it takes a `std::string`. It is not taking binary data or anything like that.

Reply:
Thanks, that snippet clarifies the intent a lot